rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,441 @@
+containing a single block of A/B. This would require the user to call
+`dnnl::memory::set_data_handle()` on each of these objects before
+brgemm call. It would incur some unnecessary overhead, but should
+not be anything substantial.
+
+However, one might wonder why use memory descriptors at all and not
+just take pointers directly as input to brgemm::execute
+call. Internally we support multiple flavors:
+- array of pointers: here the user will have to do pointer arithmetic
+  before passing those pointers to brgemm::execute call.
+- base pointer + arrray of offsets: here the pointer arithmetic will
+  be handled internally behind brgemm::execute call.
+- base pointer + fixed stride: this flavor has limited application and
+  is typically used for matmul implementation. Given its limited
+  applicability, it seems unnecessary to expose it.
+  
+To make the public brgemm API simpler, we propose to expose only the array
+of pointers flavor.
+
+For the brgemm API specifically, an array of equal sizes for A and B
+blocks are passed.  To guarentee these are of equal size, and simplify
+pointer arithmetic logic inside generated kernels, we will take an
+array of pairs of pointers.
+
+### Support for hardware-specific layouts
+
+To speedup computation for low accuracy datatype, various hardware
+require the data to be packed in a certain way:
+- grouped by 32-bit chunks along K dimension (e.g. Intel VNNI/AMX, ARM
+  SVE BFDOT). This typically stems from the larger output datatype for
+  these instructions (int32 or float).
+- grouped by 64-bit chunks along K (e.g. ARM *MMLA extensions). This
+  stems from the registers actually holding 2d matrices.
+
+Exposing these packed layouts is necessary to benefit from HW
+acceleration. This packing can happen in one of two places:
+- at execution time upon calls to brgemm::execute.
+- ahead of execution time (pre-packing) when one of the matrix is
+  constant.
+
+#### How to express HW specific layouts
+
+There are two main ways oneDNN can expose those layouts:
+- through layout tags. This has the advantage of being very simple,
+  but comes with two limitations:
+    - These tags can contain only information about data layout, not
+      extra metadata (e.g. zero-point related)
+	- it does not scale well when the number of layouts increases.
+- through memory descriptors. This can be more tedious to use as it is
+  an opaque structure that needs to be queried for
+  information. However, it is very flexible as it allows to express
+  non trivial layout descritions, and can be extended with extra
+  information/meta-data.
+
+
+Layout tags here will have to be different than what is used for the
+oneDNN primitive API, as they would mostly map to part of the
+innerblock component of existing memory::desc objects.
+
+The recommendation here would be to go with layout tags only if no
+handling of metadata and extra information is expected from oneDNN
+block level APIs.
```

Comment:
This is a very good question and was actually the initial plan as it should be much simpler to have `brgemm::packA(void *src, void *dst, bool transpose)` (and similar `brgemm::packB`). The main reasons I did not keep that was:
- when no packing is necessary, it would still require the user to make a copy in case a copy is needed (unless we expose a `brgemm::need_packA()` method);
- when packing happens independently of brgemm, this would require to jit a full brgemm object
- when user just needs to do a simple transposition, it is awkward to have to go through a brgemm object.

So the solution was to expose the simple layouts, so that user do not have to jit brgemm kernel to just do pre-packing.
But if pre-packing independently of brgemm turns out to not be required, I can add the option with packA/packB model.

Reply:
> * when no packing is necessary, it would still require the user to make a copy in case a copy is needed (unless we expose a `brgemm::need_packA()` method);

Yes I agree that if packing functions were added to the API, we would need a mechanism to avoid a unneeded copy. One option would be to have a function that returns the necessary buffer size to the packed buffers. If size is zero, then no buffer is needed and $A_i$ / $B_i$ can be passed directly to brgemm kernel.

> * when user just needs to do a simple transposition, it is awkward to have to go through a brgemm object.

Perhaps the packing routines should be responsible to prepare data to compute kernel (contiguous buffer, transpose if requested, potential alpha scaling or data conversion). Nevertheless, I understand that transposition by itself is useful feature to have.

> But if pre-packing independently of brgemm turns out to not be required, I can add the option with packA/packB model (and remove the desc part to simplify).

If prepacking independently is needed, I don't have a good suggestion.

I was think packA/packB as aiding functions for compute brgemm kernel rather than utility functions to the user for other tasks. The advantage of packing routines is that it can add some flexibility to the implementation (of brgemm), but probably we shouldn't discuss those for API requirements.

It should be fine to resolve this, since the current implementation doesn't current provide packing functions (as part of brgemm). Also, it is probably a better idea to have some version of API out such users can try and provide feedback.

Reply:
I prefer to let the user select the layout by passing layout tag, such as AB2a or AB4a because We want to generate the packA/packB kernel by our compiler as we can further fold/fuse it.